pkg/util/containers/v2/metrics/provider/collector_cache.go

```diff
@@ -0,0 +1,128 @@
+// this cache is fully synchronized to minimize locking. If a method is called multiple times in parallel for the same cachey key,
+// it may result in multiple calls to the underlying collector
+type collectorCache struct {
+	collector   Collector
+	cache       map[string]cacheEntry
+	cacheLock   sync.RWMutex
+	gcTimestamp time.Time
+}
+
+// NewCollectorCache returns a new cache dedicated to a collector
+func NewCollectorCache(collector Collector) Collector {
+	return &collectorCache{
+		collector: collector,
+		cache:     make(map[string]cacheEntry),
+	}
+}
+
+// ID returns the actual collector ID, no cache
+func (cc *collectorCache) ID() string {
+	return cc.collector.ID()
+}
+
+// GetContainerStats returns the stats if in cache and within cacheValidity
+// errors are cached as well to avoid hammering underlying collector
+func (cc *collectorCache) GetContainerStats(containerID string, cacheValidity time.Duration) (*ContainerStats, error) {
+	currentTime := time.Now()
+	cacheKey := contStatsCachePrefix + containerID
+	entry, found, err := cc.getCacheEntry(currentTime, cacheKey, cacheValidity)
+	if found {
+		if err != nil {
+			return nil, err
+		}
+
+		return entry.(*ContainerStats), nil
+	}
+
+	// No cache or wrong cache value
+	cstats, err := cc.collector.GetContainerStats(containerID, cacheValidity)
+	if err != nil {
+		cc.storeCacheEntry(currentTime, cacheKey, nil, err)
+		return nil, err
+	}
+
+	cc.storeCacheEntry(currentTime, cacheKey, cstats, nil)
+	return cstats, nil
+}
+
+// GetContainerNetworkStats returns the stats if in cache and within cacheValidity
+// errors are cached as well to avoid hammering underlying collector
+func (cc *collectorCache) GetContainerNetworkStats(containerID string, cacheValidity time.Duration) (*ContainerNetworkStats, error) {
+	// Generics could be useful. Meanwhile copy-paste.
+	currentTime := time.Now()
+	cacheKey := contNetStatsCachePrefix + containerID
+	entry, found, err := cc.getCacheEntry(currentTime, cacheKey, cacheValidity)
+	if found {
+		if err != nil {
+			return nil, err
+		}
+
+		return entry.(*ContainerNetworkStats), nil
+	}
+
+	// No cache or wrong cache value
+	val, err := cc.collector.GetContainerNetworkStats(containerID, cacheValidity)
+	if err != nil {
+		cc.storeCacheEntry(currentTime, cacheKey, nil, err)
+		return nil, err
+	}
+
+	cc.storeCacheEntry(currentTime, cacheKey, val, nil)
+	return val, nil
+}
+
+func (cc *collectorCache) getCacheEntry(currentTime time.Time, key string, cacheValidity time.Duration) (interface{}, bool, error) {
+	cc.cacheLock.RLock()
+	entry, found := cc.cache[key]
+	cc.cacheLock.RUnlock()
+
+	if !found {
+		return nil, false, nil
+	}
+
+	if currentTime.Sub(entry.timestamp) > cacheValidity {
+		return nil, false, nil
+	}
+
+	if entry.err != nil {
+		return nil, true, entry.err
+	}
+
+	return entry.value, true, nil
+}
+
+func (cc *collectorCache) storeCacheEntry(currentTime time.Time, key string, value interface{}, err error) {
+	cc.cacheLock.Lock()
```

Comment:
I'd use `defer` to unlock here to avoid forgetting about it, but it's just a personal preference.

Reply:
Indeed, it this case `defer` should be used